internal/services/keyvault/key_vault_secret_versions_data_source.go

```diff
@@ -0,0 +1,216 @@
+	}
+}
+
+func (r KeyVaultSecretVersionsDataSource) Attributes() map[string]*schema.Schema {
+	return map[string]*schema.Schema{
+		"versions": {
+			Type:     schema.TypeList,
+			Computed: true,
+			Elem: &schema.Resource{
+				Schema: map[string]*schema.Schema{
+					"id": {
+						Type:     schema.TypeString,
+						Computed: true,
+					},
+
+					"created_date": {
+						Type:     schema.TypeString,
+						Computed: true,
+					},
+
+					"enabled": {
+						Type:     schema.TypeBool,
+						Computed: true,
+					},
+
+					"not_before_date": {
+						Type:     schema.TypeString,
+						Computed: true,
+					},
+
+					"expiration_date": {
+						Type:     schema.TypeString,
+						Computed: true,
+					},
+
+					"updated_date": {
+						Type:     schema.TypeString,
+						Computed: true,
+					},
+					"uri": {
+						Type:     schema.TypeString,
+						Computed: true,
+					},
+				},
+			},
+		},
+	}
+}
+
+func (r KeyVaultSecretVersionsDataSource) ResourceType() string {
+	return "azurerm_key_vault_secret_versions"
+}
+
+func (r KeyVaultSecretVersionsDataSource) ModelObject() interface{} {
+	return &KeyVaultSecretVersionsDataSourceModel{}
+}
+
+func (r KeyVaultSecretVersionsDataSource) Read() sdk.ResourceFunc {
+	return sdk.ResourceFunc{
+		Timeout: 5 * time.Minute,
+
+		Func: func(ctx context.Context, metadata sdk.ResourceMetaData) error {
+			client := metadata.Client.KeyVault.ManagementClient
+
+			var model KeyVaultSecretVersionsDataSourceModel
+
+			if err := metadata.Decode(&model); err != nil {
+				return fmt.Errorf("decoding: %+v", err)
+			}
+
+			keyVaultId, err := commonids.ParseKeyVaultID(model.KeyVaultId)
+			if err != nil {
+				return err
+			}
+
+			keyVaultUri, err := metadata.Client.KeyVault.BaseUriForKeyVault(ctx, *keyVaultId)
+			if err != nil {
+				return err
+			}
+
+			maxResults32 := int32(model.MaxResults)
+
+			resp, err := client.GetSecretVersions(ctx, *keyVaultUri, model.Name, &maxResults32)
+			if err != nil {
+				return fmt.Errorf("making List Versions request on Azure KeyVault Secret %s: %+v", model.Name, err)
+			}
+
+			if resp.Values() != nil {
+				for resp.NotDone() {
+					for _, v := range resp.Values() {
+						model.Versions = append(model.Versions, expandSecretVersion(&v))
+					}
+					err = resp.NextWithContext(ctx)
+					if err != nil {
+						return fmt.Errorf("iterating over Secret Versions: %+v", err)
+					}
+				}
+
+				var errors []error
+				model.Versions, errors = sortSecretVersions(model.Versions)
+				if len(errors) > 0 {
+					return fmt.Errorf("sorting Secret Versions: %+v", errors)
+				}
+			}
+
+			metadata.ResourceData.SetId(fmt.Sprintf("%s/%s", model.KeyVaultId, model.Name))
+
+			return metadata.Encode(&model)
+		},
+	}
+}
+
+func expandSecretVersion(v *keyvault.SecretItem) secretVersionModel {
+	var item secretVersionModel
+	item.Uri = *v.ID
+	item.ID = (*v.ID)[strings.LastIndex(*v.ID, "/")+1:]
+	item.CreatedDate = time.Time(*v.Attributes.Created).Format(time.RFC3339)
```

Comment:
We're nil checking the other dates/times below before casting, I think we should do that here as well 
```suggestion
	createdDate := v.Attributes.Created; createdDate != nil {
		item.CreatedDate = time.Time(*createdDate).Format(time.RFC3339)
	}
```